tests/core/test_providers_manager.py

```diff
@@ -179,3 +188,13 @@ def test_auth_backends(self):
         provider_manager = ProvidersManager()
         auth_backend_module_names = list(provider_manager.auth_backend_module_names)
         assert len(auth_backend_module_names) > 0
+
+    @patch("builtins.__import__", new=raising_import)
+    def test_optional_features(self):
+        with self._caplog.at_level(logging.WARNING):
+            provider_manager = ProvidersManager()
+            list(provider_manager.field_behaviours)
+            # exasol raises regular import error with warning
+            assert "Custom Import Error: pyexasol" in str(self._caplog.messages)
+            # leveldb raises the AirflowOptionalProviderFeatureException with no warning
+            assert "Custom Import Error: plyvel" not in str(self._caplog.messages)
```

Comment:
This doesn’t feel like a good way to test the feature, it depends too much on how individual providers are set up. Maybe it’s better to dynamically create a fake provider that does the right thing on import instead.

Reply:
Yeah. Crossed my mind too. Might be to much of a shortcut indeed.

Reply:
Right. I think I have better/true unit test cases - also while testing it, I found that we can improve the import a little bit -rather than importing  Hooks once in sanity check and second time in _import_hooks, we import it only once in the _sanity_check and use imported Hook class after + I fixed class methods definition good for MyPy.